app/src/main/java/org/mozilla/fenix/share/SaveToPDFMiddleware.kt

```diff
@@ -0,0 +1,43 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.share
+
+import android.content.Context
+import mozilla.components.browser.state.action.BrowserAction
+import mozilla.components.browser.state.action.EngineAction
+import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.lib.state.Action
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+import mozilla.components.ui.tabcounter.TabCounter
+import org.mozilla.fenix.R
+import org.mozilla.fenix.components.FenixSnackbar
+import org.mozilla.fenix.components.FenixSnackbar.Companion.LENGTH_LONG
+import org.mozilla.fenix.ext.getRootView
+
+/**
+ * [BrowserAction] middleware reacting in response to Save to PDF related [Action]s.
+ * @property context The [TabCounter] that will be setup with event handlers.
+ */
+class SaveToPDFMiddleware(
+    private val context: Context,
+) : Middleware<BrowserState, BrowserAction> {
+    override fun invoke(
+        ctx: MiddlewareContext<BrowserState, BrowserAction>,
+        next: (BrowserAction) -> Unit,
+        action: BrowserAction,
+    ) {
+        if (action is EngineAction.SaveToPdfErrorAction) {
+            val rootView = context.getRootView() ?: return
+            FenixSnackbar.make(
+                view = rootView,
+                duration = LENGTH_LONG,
+                isDisplayedWithBrowserToolbar = true,
+            ).setText(context.getString(R.string.unable_to_save_to_pdf_error))
+                .show()
+        }
+        next(action)
```

Comment:
Do we also want to report telemetry to know how many times Save to PDF failed @kbrosnan ? 

Value: We have telemetry for `save_to_pdf_tapped` events, so we could better observe the rate of failures as the feature goes out? We can let this expire after 2-3 releases which would be useful platform telemetry to have.

Reply:
This will be address in separated ticket